numpyro/util.py

```diff
@@ -270,14 +275,362 @@ def update_fn(t, accept_prob, z_flat, state):
+                                               index_update(x[1], ckpt_idx_max, r_sum)))
+
+    def _body_fn(state):
+        i, _ = state
+        subtree_r_sum = r_sum - r_sum_ckpts[i] + r_ckpts[i]
+        # XXX we might not need to unravel here
+        return i - 1, _is_turning(inverse_mass_matrix, r_ckpts[i], r, subtree_r_sum)
+
+    _, turning = lax.while_loop(lambda it: (it[0] >= ckpt_idx_min) & ~it[1],
+                                _body_fn,
+                                (ckpt_idx_max, False))
+    return turning, r_ckpts, r_sum_ckpts
+
+
+def _iterative_build_subtree(depth, vv_update, kinetic_fn, z, r, z_grad,
+                             inverse_mass_matrix, step_size, going_right, rng,
+                             energy_current, slice_exp_term, max_sliced_energy,
+                             use_multinomial_sampling, max_tree_depth):
+    max_num_proposals = 2 ** depth
+
+    def _cond_fn(state):
+        tree, turning, _, _, _ = state
+        return (tree.num_proposals < max_num_proposals) & ~turning & ~tree.diverging
+
+    def _body_fn(state):
+        current_tree, _, r_checkpoints, r_sum_checkpoints, rng = state
+        rng, transition_rng = random.split(rng)
+        z, r, z_grad = _get_leaf(current_tree, going_right)
+        new_leaf = _build_basetree(vv_update, kinetic_fn, z, r, z_grad, step_size, going_right,
+                                   energy_current, slice_exp_term, max_sliced_energy,
+                                   use_multinomial_sampling)
+        new_tree = _combine_tree(current_tree, new_leaf, inverse_mass_matrix, going_right,
+                                 transition_rng, use_multinomial_sampling,
+                                 _uniform_transition_prob, iterative_build=True)
+        turning, r_checkpoints, r_sum_checkpoints = _is_iterative_turning(
+            current_tree.num_proposals,
+            inverse_mass_matrix,
+            new_leaf.r_right,
+            new_tree.r_sum,
+            r_checkpoints,
+            r_sum_checkpoints
+        )
+        return new_tree, turning, r_checkpoints, r_sum_checkpoints, rng
+
+    basetree = _build_basetree(vv_update, kinetic_fn, z, r, z_grad, step_size, going_right,
+                               energy_current, slice_exp_term, max_sliced_energy,
+                               use_multinomial_sampling)
+    r_init, _ = ravel_pytree(basetree.r_left)
+    # TODO: we can create these checkpoints at build_tree method
+    # and reuse it; but let's do this optimization later
+    r_checkpoints = np.zeros((max_tree_depth, inverse_mass_matrix.shape[-1]),
+                             dtype=inverse_mass_matrix.dtype)
+    r_checkpoints = index_update(r_checkpoints, 0, r_init)
+    r_sum_checkpoints = np.zeros((max_tree_depth, inverse_mass_matrix.shape[-1]),
+                                 dtype=inverse_mass_matrix.dtype)
+    r_sum_checkpoints = index_update(r_sum_checkpoints, 0, r_init)
+
+    # FIXME: use native while here is fine
```

Comment:
iterative program works correctly now; the only thing which blocks us from jitting NUTS is this while_loop; not sure why we get **RuntimeError: Invalid argument: The parameter of condition and body, the result of the body, and init must all have the same shape; got Condition: ...** with JAX here while the native python `while` loop works. will investigate more